Content.Client/Visualizer/ToggleVisualizer.cs

```diff
@@ -0,0 +1,59 @@
+﻿using JetBrains.Annotations;
+using Robust.Client.GameObjects;
+using Robust.Shared.Serialization.Manager.Attributes;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Client.Visualizer
+{
+    /// <summary>
+    /// General purpose appearance visualizer used to toggle back and forth between two states
+    /// </summary>
+    [UsedImplicitly]
+    public class ToggleVisualizer : AppearanceVisualizer
+    {
+        /// <summary>
+        /// The suffix used to name the state. Visualizer switches between [state] and [state]_[suffix]
+        /// For example "object", "object_enabled"
+        /// </summary>
+        [DataField("state_suffix", required: true)]
+        private string? _suffix;
+
+        /// <summary>
+        /// The key the visualizer uses to get toggled state data
+        /// </summary>
+        [DataField("key", required: false)]
+        private string _key = "Toggle";
+
+        public override void OnChangeData(AppearanceComponent appearance)
+        {
+            base.OnChangeData(appearance);
+
+            if (!appearance.Owner.TryGetComponent(out SpriteComponent? sprite)) return;
+
+            if (!appearance.TryGetData(_key, out bool toggle)) return;
+
+            // Set all the layers state to [state]_[suffix] if it exists, or back to the default [state]
+            // If they don't exist, leave them like that
+            var i = 0;
+
+            foreach (var layer in sprite.AllLayers)
+            {
+                if (string.IsNullOrEmpty(_suffix))
+                    continue;
+
+                var newState = toggle
+                    ? sprite.LayerGetState(i) + _suffix
+                    : sprite.LayerGetState(i).ToString()?.Replace(_suffix, "");
+
+                // Check if the state actually exists in the RSI and set it on the layer
+                var stateExists = sprite.BaseRSI?.TryGetState(newState, out var actualState);
+                if (stateExists ?? false)
+                {
+                    sprite.LayerSetState(i, newState);
+                }
+
+                i++;
+            }
```

Comment:
Can't you just have explicit buckle layers? Iterating every layer is pretty dirty. You're also doing the O(1) check in the loop...

Reply:
As stated in another comment, this is to make the StateVisualizer flexible, having to specify explicitly every layer on every prototype that will use it would be a maintaining nightmare. 
With this you just have to create a new state with the suffix for each layer _and_ direction and the visualizer will figure it out. It's not that costly also

Reply:
Specifying 1 layer or assuming one doesn't seem that painful, you already have to specify the suffix anyway.